### DISCUSSION TITLE
コマンド実行のインターフェイス

### DISCUSSION THREAD

**Discussion by torus** [torus/42-minishell#17] [Ideas]
パーサからコマンド実行へ情報を渡すとき、AST をそのまま渡すと無駄な情報が含まれてしまいデータ構造も無駄に複雑になるので、必要十分な情報にしぼって受け渡しするほうが望ましいかと思います。

たとえば、コマンドラインでは `command > file1 > file2 >> file3` などのように標準出力のリダイレクト先を複数指定できるけど、実際に有効になるのは最後に指定したものだけなので、file1 や file2 の処理はコマンド実行の前段階で処理できます。（具体的にはファイルをオープンしてトランケートするだけ。）

で、いま考えているのが、こんなデータ構造です。

```c
typedef struct s_command_invocation
{
    const char              *output_file_path; /* > file の場合 */
    t_command_invocation    *piped_command; /* | command の場合（> と | は同時には指定できない） */
    const char              *input_file_path; /* < file */
    const char              **exec_and_args; /* ft_split で区切ったコマンドのパスと引数のリスト */
}   t_command_invocation;
```

複数コマンドが `;` で区切られている場合は、単にコマンド実行を逐次的に呼び出せばいいだけだと思うので、まずはパイプでつながったコマンドを実行するところから始めるのが良さそうな気がします。

とりあえず取っ掛かりとしてこんなのでどうでしょうか？

**Comment by torus** [comment]
```
$ echo hello > dog.txt | wc
      0       0       0
$ 
```
こういうパターンは別途考える必要があるな。

この場合、wc コマンドを起動して、wc に空の入力ストリームを渡す感じかな。

**Reply by torus** [reply]
上のコマンドは `echo hello > dog.txt ; wc < /dev/null` に置き換えられるな。

**Reply by torus** [reply]
あ、厳密には wc は echo と同時に起動されなきゃいけないから、ちょっと違う。

**Comment by JUNNETWORKS** [comment]
これ良いと思います!
一度これでCommandExecutionを実装してみます!
